apnsworker/src/main/scala/apnsworker/Apns.scala

```diff
@@ -0,0 +1,42 @@
+package apnsworker
+
+import cats.effect._
+import cats.syntax.functor._
+import com.turo.pushy.apns.{ApnsClient => PushyApnsClient}
+import db.{RegistrationService, ShardRange, Topic}
+import _root_.models.{Notification, iOS}
+import fs2.Stream
+import models.ApnsConfig
+
+import scala.concurrent.{ExecutionContext, Future, blocking}
+
+class Apns[F[_]](registrationService: RegistrationService[F, Stream], config: ApnsConfig)
+  (implicit executionContext: ExecutionContext, contextShift: Concurrent[F], F: Async[F]) {
+
+  private val apnsClient: F[PushyApnsClient] =
+    ApnsClient(config).fold(e => F.raiseError(e), c => F.delay(c))
+
+  def send(notification: Notification, shardRange: ShardRange): Stream[F, Either[Throwable, String]] = {
+
+    def sendAsync(token: String)(client: PushyApnsClient): F[String] =
+      Async[F].async[String] { (cb: Either[Throwable, String] => Unit) =>
+        ApnsClient.sendNotification(token, notification)(cb)(client, config)
+      }
+
+    def tokens: Stream[F, String] = notification
+        .topic
+        .map(topic => registrationService.find(Topic(topic.name), iOS, shardRange))
+        .reduce(_ merge _)
+        .map(_.device.token)
+
+    for {
+      client <- Stream.eval(apnsClient)
+      res <- tokens
+        .map(token => Stream.eval(sendAsync(token)(client)).attempt)
+        .parJoinUnbounded
```

Comment:
what's `parJoinUnbounded` doing?

Reply:
We have a `Stream` of `Stream` (`Stream[Stream[ResponseEither]]`). The outer stream emitting `tokens` and the inner stream emiting the `send response` for a given token.
`parJoinUnbounded` merge the `Stream` of `Stream` into a single `Stream` (`Stream[ResponseEither]`) running all inner `Streams` at the same time.
I am using `parJoinUnbounded` rather than `parJoin(maxConcurrenty: Int)` because `Pushy` handles for us the limitation of in-flight requests.

Reply:
got it :+1: